src/VisualStudio/Core/Def/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -221,6 +282,45 @@ public int FormatSpan(IVsTextLines pBuffer, VsTextSpan[] tsInSurfaceBuffer)
 
             var formattingSpan = CommonFormattingHelpers.GetFormattingSpan(SubjectBuffer.CurrentSnapshot, snippetTrackingSpan.GetSpan(SubjectBuffer.CurrentSnapshot));
 
+            // If the snippet spans more than one line and a specific line ending is expected, normalize the line
+            // endings inside the snippet span before continuing.
+            if (newLineEnding is not null && tsInSurfaceBuffer[0].iEndLine > tsInSurfaceBuffer[0].iStartLine)
+            {
+                var totalOffset = 0;
+                var changes = new List<TextChange>();
+                var existingText = SubjectBuffer.CurrentSnapshot.GetText(formattingSpan.ToSpan());
+                for (var i = 0; i < existingText.Length; i++)
+                {
+                    var nextLineEnding = existingText.IndexOfAny(['\r', '\n'], i);
+                    if (nextLineEnding < 0)
+                        break;
+
+                    i = nextLineEnding;
+                    if (existingText[i] == '\r' && i < existingText.Length - 1 && existingText[i + 1] == '\n')
+                    {
+                        if (newLineEnding != "\r\n")
+                        {
+                            changes.Add(new TextChange(new TextSpan(formattingSpan.Start + i, 2), newLineEnding));
+                            totalOffset += newLineEnding.Length - 2;
+                        }
+
+                        // Move to the \n, which is the last character of the currently-located line ending
+                        i++;
+                    }
+                    else if (existingText[i] is '\r' or '\n')
+                    {
+                        if (newLineEnding.Length > 1 || newLineEnding[0] != existingText[i])
+                        {
+                            changes.Add(new TextChange(new TextSpan(formattingSpan.Start + i, 2), newLineEnding));
+                            totalOffset += newLineEnding.Length - 1;
+                        }
+                    }
+                }
```

Comment:
Since we've already using the editor APIs, just looping through the lines collection and using the spans already identified as the linebreak spans whould be much simpler, as well as removing the GetText() which might realize a long span.

Reply:
Will investigation this option.